### PR TITLE
Add limitsFromChannel property to PyDMSpinbox

### PR DESCRIPTION
**Pull request by fernandohds564**
In some cases PVs do not come with its control limits properly set. In these cases the limits of the widget are set to zero and it is not possible to use it to change the PV values.
This PR tries to address this issue by creating a `PyQt` `Property` called `limitsFromChannel` which allows the user to choose if the widget limit should follow the values defined by the channel.

### CODE REVIEW

File: pydm/tests/widgets/test_spinbox.py [hunk 1]
```diff
@@ -82,49 +90,70 @@ def test_key_press_event(qtbot, signals, monkeypatch, first_key_pressed, second_
...
 
-    def press_key_and_verify(key_pressed, key_mod, key_press_count, expected_exp, expected_value):
+    def press_key_and_verify(
+            key_pressed, key_mod, key_press_count, expected_exp,
+            expected_value):
         if key_mod != Qt.NoModifier:
-            # Monkeypatch the Control flag because in this test, we don't properly have the app QApplication instance
-            monkeypatch.setattr(QApplication, "queryKeyboardModifiers", lambda *args: key_mod)
+            # Monkeypatch the Control flag because in this test, we don't
+            # properly have the app QApplication instance
```
Comment (hhslepicka) [inline pydm/tests/widgets/test_spinbox.py]: You can have the qapp by just adding it to the list of parameters by the side of the qtbot... it is a fixture of PyTest...

File: pydm/tests/widgets/test_spinbox.py [hunk 2]
```diff
@@ -270,50 +307,73 @@ def test_send_value(qtbot, signals, init_value, user_typed_value, precision):
...
+        "UPPER" if the new value is intended for the upper limit, "LOWER" for
+        the lower limit
     new_limit : float
         The new limit value
+    from_pv : bool
+        If True the PyDMSpinbox limits will follow the ctrl limits from PV.
     """
     pydm_spinbox = PyDMSpinbox()
     qtbot.addWidget(pydm_spinbox)
+    pydm_spinbox.limitsFromPV = from_pv
```
Comment (hhslepicka) [inline pydm/tests/widgets/test_spinbox.py]: `limitsFromPV` -> `limitsFromChannel`

File: pydm/widgets/spinbox.py [hunk 3]
```diff
@@ -21,6 +21,7 @@ def __init__(self, parent=None, init_channel=None):
         self.setEnabled(False)
         self._alarm_sensitive_border = False
         self._show_step_exponent = True
+        self._limits_from_pv = True
```
Comment (hhslepicka) [inline pydm/widgets/spinbox.py]: We are trying to not use the name `PV` anymore... `Channel` is the proper name for PyDM. Do you mind to change: `_limits_from_pv` -> `_limits_from_channel`

File: pydm/widgets/spinbox.py [hunk 4]
```diff
@@ -175,3 +182,34 @@ def showStepExponent(self, val):
         """
         self._show_step_exponent = val
         self.update_format_string()
+
+    @Property(bool)
+    def limitsFromPV(self):
```
Comment (hhslepicka) [inline pydm/widgets/spinbox.py]: `limitsFromPV` -> `limitsFromChannel`

File: pydm/widgets/spinbox.py [hunk 5]
```diff
@@ -175,3 +182,34 @@ def showStepExponent(self, val):
...
+        Returns
+        -------
+        limits_from_pv : bool
+            True means that the widget will use the limits information
+            from the Channel if available.
+        """
+        return self._limits_from_pv
+
+    @limitsFromPV.setter
+    def limitsFromPV(self, value):
```
Comment (hhslepicka) [inline pydm/widgets/spinbox.py]: `limitsFromPV` -> `limitsFromChannel`

File: pydm/tests/widgets/test_spinbox.py (hunk 1 above)
Comment (fernandohds564) [reply pydm/tests/widgets/test_spinbox.py]: Hi @hhslepicka. I just fixed the style here splitting the comment in two lines... I didn't change its content.

**Comment by fernandohds564** [comment]
@hhslepicka, I have noticed that `PyDMSlider` have properties with similar funcionality to the ones implemented in `PyDMScale` but different names: `userDefinedLimits`, `userMaximum` and `userMinimum`. Do you think it is a good idea to rename them?

**Comment by hhslepicka** [comment]
@fernandohds564, in the case of the limits it is all good. I forgot about the setMin and Max at the QDoubleSpinBox.

For the PyDMSlider it is better to leave as is for now as we don’t want to break screens deployed in the wild. We can later think about a way to deprecate the properties and add new ones to be uniform.

**Comment by fernandohds564** [comment]
@hhslepicka, in this meantime I implemented the properties userUpperLimit and userLowerLimit. I think it you were right, it is better this way. What do you think?

File: pydm/tests/widgets/test_spinbox.py (hunk 1 above)
Comment (fernandohds564) [reply pydm/tests/widgets/test_spinbox.py]: @hhslepicka, can I mark this one as resolved?

File: pydm/widgets/spinbox.py [hunk 6]
```diff
@@ -175,3 +201,79 @@ def showStepExponent(self, val):
...
+        """
+        The user-defined lower limit for the spinbox.
+        Parameters
+        ----------
+        value : float
+            The new lower limit value.
+        """
+        self._user_lower_limit = float(value)
+        self._user_upper_limit = max(
+            self._user_upper_limit, self._user_lower_limit)
```
Comment (fernandohds564) [inline pydm/widgets/spinbox.py]: @hhslepicka, I would like to know your opinion about two points here:
 1) I always set the attributes _user_lower_limit regardless of the state of _limits_from_channel. I noticed that this is not being done at PyDMSlider and PyDMScale widgets. Do you think this is appropriate?
 2) I check if the lower value being set is larger than the current maximum and move the maximum to leave the range valid. I saw that this is what QSpinBox does. I also noticed that this is not done in PyDMSlider and PyDMScale.

File: pydm/widgets/spinbox.py [hunk 7]
```diff
@@ -175,3 +201,79 @@ def showStepExponent(self, val):
...
+        The user-defined upper limit for the spinbox.
+        Parameters
+        ----------
+        value : float
+            The new upper limit value.
+        """
+        self._user_upper_limit = float(value)
+        self._user_lower_limit = min(
+            self._user_lower_limit, self._user_upper_limit)
+        self.update_limits()
```
Comment (fernandohds564) [inline pydm/widgets/spinbox.py]: The same comments for the `userLowerLimit` setter applies here.

**Comment by fernandohds564** [comment]
Hi @hhslepicka, during the weekend I had another idea. Now you have three options of implementations to choose:

 1) The original implementation, without the properties for the user to set minimum and maximum. In this case the user must define the limits using `setMinimum`, `setMaximum` and `setRange` methods of the widget and they are not stored internally, in such a way that, if `limitsFromChannel` is set to True, they will be lost when connection is established.

 2) commit ff2ccad : In this case, there are two properties `userUpperLimit` and `userLowerLimit`, which can be used by the user to set the desired limits. If `limitsFromChannel` is set to `True`, the values given by the user are stored and are not lost during the connections with the channel. This way, if at run time the user changes the value of `limitsFromChannel`, they will be restored. The disadvantage of this option is that the methods `setMinimum` and `setMaximum` of the parent widget are still available, and consistency is not guaranteed if they are used.

 3) The HEAD of this branch (actually I don't know why the head showing here was not updated yet, I already have pushed my new commits): I removed the `userUpperLimit` and `userLowerLimit` and reimplemented `setMiminum`, `setMaximum` and `setRange`. With this implementation there is no easy way to break the interface or create inconsistencies. The action of setting the properties `minimum` and `maximum` of the parent widget reflect the user defined limits and its application to the real widget limits is susceptible to the the state of `limitsFromChannel`. The disadvantage of this option is that the names of the properties do not explicitly expose their scope of validity and the value retrieved from a call to `minimum()` and `maximum()` will not reflect the user defined limits, but the current valid limit of the widget. The user defined values are accessible by `get_user_limits()` method. 

Which one do you prefer? Do you have other suggestions and ideas? this problem seemed simple at first but there is a lot of subtleties...

**Pull request closed without merging**